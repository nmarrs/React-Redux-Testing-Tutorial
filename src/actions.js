import { templateTasksRef } from './config/firebase';

/**
  Action function that fetches data from firebase database
*/
export const fetchData = () => async dispatch => {
  templateTasksRef.on('value', snapshot => {
    dispatch({
      type: 'FETCH_TEMPLATE_TASKS',
      payload: snapshot.val()
    })
  })
}

/**
  Action function that adds new template task to firebase database
*/
export const addTemplateTask = (newTemplateTask) => async dispatch => {
  templateTasksRef.push().set(newTemplateTask)
}

/**
  Action function that edits template task in firebase database
*/
export const editTemplateTask = (templateTask, templateTaskId) => async dispatch => {
  templateTasksRef.child(templateTaskId).update(templateTask)
}

/**
  Action function that opens / closes the create task modal
*/
export const toggleCreateTaskModal = (showCreateTaskModal) => {
  return {
    type: 'TOGGLE_CREATE_TASK_MODAL',
    payload: showCreateTaskModal
  }
}

/**
  Action function that opens / closes the edit task modal
*/
export const toggleEditTaskModal = (showEditTaskModal, task) => {
  return {
    type: 'TOGGLE_EDIT_TASK_MODAL',
    payload: showEditTaskModal,
    task: task
  }
}

/**
  Action function that toggles the admin / user view
*/
export const toggleIsAdminView = () => {
  return {
    type: 'TOGGLE_IS_ADMIN_VIEW'
  }
}

/**
  Action function that assigns the current template tasks to the user
*/
export const assignTasksToUser = (templateTasks) => {
  window.alert('Tasks have been assigned to user!')
  return {
    type: 'ASSIGN_TASKS_TO_USER',
    payload: templateTasks
  }
}

/**
  Action function that creates a new task (template or regular)
*/
export const createTask = (task, isAdminView) => {
  if (isAdminView) {
    return {
      type: 'CREATE_TEMPLATE_TASK',
      payload: task
    }
  } else {
    return {
      type: 'CREATE_TASK',
      payload: task
    }
  }
}

/**
  Action function that edits a task (template or regular)
*/
export const editTask = (task, isAdminView) => {
  if (isAdminView) {
    return {
      type: 'EDIT_TEMPLATE_TASK',
      payload: task
    }
  } else {
    return {
      type: 'EDIT_TASK',
      payload: task
    }
  }
}

/**
  Action function that sorts application task by rank weight

  Tasks are sorted to list the most prioritized and shortest tasks first and the
  least prioritized and longest tasks last.
*/
export const sortTasksByRank = (templateTasks, userTasks) => {
  let sortedTemplateTasks = templateTasks.sort((a, b) => {
    return parseInt(b.rankWeight - a.rankWeight)
  })
  let sortedUserTasks = userTasks.sort((a, b) => {
    return parseInt(b.rankWeight - a.rankWeight)
  })

  return {
    type: 'SORT_TASKS_BY_RANK',
    sortedTemplateTasks: sortedTemplateTasks,
    sortedUserTasks: sortedUserTasks
  }
}

/**
  Helper function that caculates weight of task for ranking purposes

  Tasks are sorted to list the most prioritized and shortest tasks first and the
  least prioritized and longest tasks last.
*/
export const calculateTaskRankWeight = (priority, timeEstimate) => {
  return parseInt((priority * 100) / timeEstimate)
}
