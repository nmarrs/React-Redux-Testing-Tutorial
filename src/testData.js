export let testTask = {
  id: 1,
  title: 'Unload merchandise',
  priority: 3,
  timeEstimate: 2,
  notes: 'Unload merchandise from the bay door. Driver usually comes by at 6 am sharp.',
  feedback: '',
  startTime: '',
  endTime: '',
  completionTime: '',
  currentStatus: 'To Do',
  rankWeight: 150
}

export let testTask1 = {
  id: 2,
  title: 'Stock shelves with unpacked items',
  priority: 1,
  timeEstimate: 10,
  notes: 'Take merchandise items and stock shelves with them. Be sure to place items in correct areas.',
  feedback: '',
  startTime: '',
  endTime: '',
  completionTime: '',
  currentStatus: 'To Do',
  rankWeight: 10
}

export let testTask2 = {
  id: 3,
  title: 'Stock shelves with unpacked items',
  priority: 5,
  timeEstimate: 2,
  notes: 'Take merchandise items and stock shelves with them. Be sure to place items in correct areas.',
  feedback: '',
  startTime: '',
  endTime: '',
  completionTime: '',
  currentStatus: 'To Do',
  rankWeight: 250
}
