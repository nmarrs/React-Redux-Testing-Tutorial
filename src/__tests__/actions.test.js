import configureStore from "redux-mock-store";
import thunk from "redux-thunk";

import * as actions from ".././actions";
import { testTask, testTask1 } from "../testData";

describe("actions", () => {
  const middlewares = [thunk];
  const mockStore = configureStore(middlewares);
  const store = mockStore();

  beforeEach(() => {
    store.clearActions();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  // INSERT TEST FOR `toggleCreateTaskModal` HERE!!

  describe("toggleEditTaskModal", () => {
    it("dispatches the correct action and payload", () => {
      const showEditTaskModal = true;

      store.dispatch(actions.toggleEditTaskModal(showEditTaskModal, testTask));
      expect(store.getActions()).toMatchSnapshot();
    });
  });

  describe("toggleIsAdminView", () => {
    it("dispatches the correct action and payload", () => {
      store.dispatch(actions.toggleIsAdminView());
      expect(store.getActions()).toMatchSnapshot();
    });
  });

  describe("assignTasksToUser", () => {
    it("dispatches the correct action and payload", () => {
      global.window.alert = jest.fn();
      let templateTasks = [testTask];

      store.dispatch(actions.assignTasksToUser(templateTasks));
      expect(store.getActions()).toMatchSnapshot();
      expect(global.window.alert).toHaveBeenCalled();
      expect(global.window.alert).toHaveBeenCalledWith(
        "Tasks have been assigned to user!"
      );
    });
  });

  describe("createTask", () => {
    it("dispatches the correct action and payload when admin view", () => {
      let isAdminView = true;

      store.dispatch(actions.createTask(testTask, isAdminView));
      expect(store.getActions()).toMatchSnapshot();
    });

    it("dispatches the correct action and payload when not admin view", () => {
      let isAdminView = false;

      store.dispatch(actions.createTask(testTask, isAdminView));
      expect(store.getActions()).toMatchSnapshot();
    });
  });

  describe("editTask", () => {
    it("dispatches the correct action and payload when admin view", () => {
      let isAdminView = true;

      store.dispatch(actions.editTask(testTask, isAdminView));
      expect(store.getActions()).toMatchSnapshot();
    });

    it("dispatches the correct action and payload when not admin view", () => {
      let isAdminView = false;

      store.dispatch(actions.editTask(testTask, isAdminView));
      expect(store.getActions()).toMatchSnapshot();
    });
  });

  describe("sortTasksByRank", () => {
    it("dispatches the correct action and payload", () => {
      let templateTasks = [testTask];
      let userTasks = [testTask, testTask1];

      store.dispatch(actions.sortTasksByRank(templateTasks, userTasks));
      expect(store.getActions()).toMatchSnapshot();
    });
  });

  describe("calculateTaskRankWeight", () => {
    it("calculates task rank weight as expected", () => {
      let priority1 = 3;
      let timeEstimate1 = 2;
      let priority2 = 1;
      let timeEstimate2 = 10;

      let taskRankWeight1 = actions.calculateTaskRankWeight(
        priority1,
        timeEstimate1
      );
      let taskRankWeight2 = actions.calculateTaskRankWeight(
        priority2,
        timeEstimate2
      );
      expect(taskRankWeight1).toMatchSnapshot();
      expect(taskRankWeight2).toMatchSnapshot();
      expect(taskRankWeight1).toBeGreaterThan(taskRankWeight2);
    });
  });
});
