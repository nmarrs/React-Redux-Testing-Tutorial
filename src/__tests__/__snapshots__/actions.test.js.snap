// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions assignTasksToUser dispatches the correct action and payload 1`] = `
Array [
  Object {
    "payload": Array [
      Object {
        "completionTime": "",
        "currentStatus": "To Do",
        "endTime": "",
        "feedback": "",
        "id": 1,
        "notes": "Unload merchandise from the bay door. Driver usually comes by at 6 am sharp.",
        "priority": 3,
        "rankWeight": 150,
        "startTime": "",
        "timeEstimate": 2,
        "title": "Unload merchandise",
      },
    ],
    "type": "ASSIGN_TASKS_TO_USER",
  },
]
`;

exports[`actions calculateTaskRankWeight calculates task rank weight as expected 1`] = `150`;

exports[`actions calculateTaskRankWeight calculates task rank weight as expected 2`] = `10`;

exports[`actions createTask dispatches the correct action and payload when admin view 1`] = `
Array [
  Object {
    "payload": Object {
      "completionTime": "",
      "currentStatus": "To Do",
      "endTime": "",
      "feedback": "",
      "id": 1,
      "notes": "Unload merchandise from the bay door. Driver usually comes by at 6 am sharp.",
      "priority": 3,
      "rankWeight": 150,
      "startTime": "",
      "timeEstimate": 2,
      "title": "Unload merchandise",
    },
    "type": "CREATE_TEMPLATE_TASK",
  },
]
`;

exports[`actions createTask dispatches the correct action and payload when not admin view 1`] = `
Array [
  Object {
    "payload": Object {
      "completionTime": "",
      "currentStatus": "To Do",
      "endTime": "",
      "feedback": "",
      "id": 1,
      "notes": "Unload merchandise from the bay door. Driver usually comes by at 6 am sharp.",
      "priority": 3,
      "rankWeight": 150,
      "startTime": "",
      "timeEstimate": 2,
      "title": "Unload merchandise",
    },
    "type": "CREATE_TASK",
  },
]
`;

exports[`actions editTask dispatches the correct action and payload when admin view 1`] = `
Array [
  Object {
    "payload": Object {
      "completionTime": "",
      "currentStatus": "To Do",
      "endTime": "",
      "feedback": "",
      "id": 1,
      "notes": "Unload merchandise from the bay door. Driver usually comes by at 6 am sharp.",
      "priority": 3,
      "rankWeight": 150,
      "startTime": "",
      "timeEstimate": 2,
      "title": "Unload merchandise",
    },
    "type": "EDIT_TEMPLATE_TASK",
  },
]
`;

exports[`actions editTask dispatches the correct action and payload when not admin view 1`] = `
Array [
  Object {
    "payload": Object {
      "completionTime": "",
      "currentStatus": "To Do",
      "endTime": "",
      "feedback": "",
      "id": 1,
      "notes": "Unload merchandise from the bay door. Driver usually comes by at 6 am sharp.",
      "priority": 3,
      "rankWeight": 150,
      "startTime": "",
      "timeEstimate": 2,
      "title": "Unload merchandise",
    },
    "type": "EDIT_TASK",
  },
]
`;

exports[`actions sortTasksByRank dispatches the correct action and payload 1`] = `
Array [
  Object {
    "sortedTemplateTasks": Array [
      Object {
        "completionTime": "",
        "currentStatus": "To Do",
        "endTime": "",
        "feedback": "",
        "id": 1,
        "notes": "Unload merchandise from the bay door. Driver usually comes by at 6 am sharp.",
        "priority": 3,
        "rankWeight": 150,
        "startTime": "",
        "timeEstimate": 2,
        "title": "Unload merchandise",
      },
    ],
    "sortedUserTasks": Array [
      Object {
        "completionTime": "",
        "currentStatus": "To Do",
        "endTime": "",
        "feedback": "",
        "id": 1,
        "notes": "Unload merchandise from the bay door. Driver usually comes by at 6 am sharp.",
        "priority": 3,
        "rankWeight": 150,
        "startTime": "",
        "timeEstimate": 2,
        "title": "Unload merchandise",
      },
      Object {
        "completionTime": "",
        "currentStatus": "To Do",
        "endTime": "",
        "feedback": "",
        "id": 2,
        "notes": "Take merchandise items and stock shelves with them. Be sure to place items in correct areas.",
        "priority": 1,
        "rankWeight": 10,
        "startTime": "",
        "timeEstimate": 10,
        "title": "Stock shelves with unpacked items",
      },
    ],
    "type": "SORT_TASKS_BY_RANK",
  },
]
`;

exports[`actions toggleCreateTaskModal dispatches the correct action and payload 1`] = `
Array [
  Object {
    "payload": true,
    "type": "TOGGLE_CREATE_TASK_MODAL",
  },
]
`;

exports[`actions toggleEditTaskModal dispatches the correct action and payload 1`] = `
Array [
  Object {
    "payload": true,
    "task": Object {
      "completionTime": "",
      "currentStatus": "To Do",
      "endTime": "",
      "feedback": "",
      "id": 1,
      "notes": "Unload merchandise from the bay door. Driver usually comes by at 6 am sharp.",
      "priority": 3,
      "rankWeight": 150,
      "startTime": "",
      "timeEstimate": 2,
      "title": "Unload merchandise",
    },
    "type": "TOGGLE_EDIT_TASK_MODAL",
  },
]
`;

exports[`actions toggleIsAdminView dispatches the correct action and payload 1`] = `
Array [
  Object {
    "type": "TOGGLE_IS_ADMIN_VIEW",
  },
]
`;
